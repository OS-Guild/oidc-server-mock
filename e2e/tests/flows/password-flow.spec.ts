import * as dotenv from 'dotenv';

import clients from '../../config/clients-configuration.json';
import users from '../../config/user-configuration.json';
import { introspectEndpoint, tokenEndpoint, userInfoEndpoint } from '../../helpers';
import type { Client, User } from '../../types';

const testCases: User[] = users
  .map(u => ({
    ...u,
    toString: function () {
      return this.SubjectId;
    },
  }))
  .sort((u1, u2) => (u1.SubjectId < u2.SubjectId ? -1 : 1));

describe('Password Flow', () => {
  let client: Client;
  let token: string;

  beforeAll(() => {
    dotenv.config();
    client = clients.find(c => c.ClientId === 'password-flow-client-id');
    expect(client).toBeDefined();
  });

  describe.each(testCases)('- %s -', (user: User) => {
    test('Token Endpoint', async () => {
      const parameters = {
        client_id: client.ClientId,
        username: user.Username,
        password: user.Password,
        grant_type: 'password',
        scope: client.AllowedScopes.join(' '),
      };

      token = await tokenEndpoint(parameters);
    });

    test('UserInfo Endpoint', async () => {
      await userInfoEndpoint(token);
    });

    test('Introspection Endpoint', async () => {
      await introspectEndpoint(token, 'some-app');
    });
  });
});
