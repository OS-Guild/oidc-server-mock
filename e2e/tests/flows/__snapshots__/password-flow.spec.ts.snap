// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Flow - simple_user - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "password-flow-client-id",
  "sub": "simple_user",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Password Flow - simple_user - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "email",
    "openid",
    "profile",
    "some-app-scope-1",
    "some-custom-identity"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "password-flow-client-id",
  "sub": "simple_user",
  "idp": "local"
}
`;

exports[`Password Flow - simple_user - UserInfo Endpoint 1`] = `
{
  "sub": "simple_user",
}
`;

exports[`Password Flow - user_with_all_claim_types - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "password-flow-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Password Flow - user_with_all_claim_types - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "email",
    "openid",
    "profile",
    "some-app-scope-1",
    "some-custom-identity"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "password-flow-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim"
}
`;

exports[`Password Flow - user_with_all_claim_types - UserInfo Endpoint 1`] = `
{
  "email": "oliver.hunter@gmail.com",
  "name": "Oliver Hunter",
  "some-custom-identity-user-claim": "Oliver's Custom User Claim",
  "sub": "user_with_all_claim_types",
}
`;

exports[`Password Flow - user_with_custom_api_resource_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "password-flow-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Password Flow - user_with_custom_api_resource_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "email",
    "openid",
    "profile",
    "some-app-scope-1",
    "some-custom-identity"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "password-flow-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim"
}
`;

exports[`Password Flow - user_with_custom_api_resource_claims - UserInfo Endpoint 1`] = `
{
  "email": "sam.tailor@gmail.com",
  "name": "Sam Tailor",
  "sub": "user_with_custom_api_resource_claims",
}
`;

exports[`Password Flow - user_with_custom_identity_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "password-flow-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Password Flow - user_with_custom_identity_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "email",
    "openid",
    "profile",
    "some-app-scope-1",
    "some-custom-identity"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "password-flow-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local"
}
`;

exports[`Password Flow - user_with_custom_identity_claims - UserInfo Endpoint 1`] = `
{
  "email": "jack.sparrow@gmail.com",
  "name": "Jack Sparrow",
  "some-custom-identity-user-claim": "Jack's Custom User Claim",
  "sub": "user_with_custom_identity_claims",
}
`;

exports[`Password Flow - user_with_standard_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "password-flow-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Password Flow - user_with_standard_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "email",
    "openid",
    "profile",
    "some-app-scope-1",
    "some-custom-identity"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "password-flow-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local"
}
`;

exports[`Password Flow - user_with_standard_claims - UserInfo Endpoint 1`] = `
{
  "email": "john.smith@gmail.com",
  "email_verified": "true",
  "name": "John Smith",
  "sub": "user_with_standard_claims",
}
`;
